### PR TITLE
adding ansible playbooks for deploying the pcn-k8s component.

### PR DESCRIPTION
**Pull request by alacuku**
The main difference between this vagrant solution and the old one is that it does the provision of the VMs with a bash script embedded in the vagrant file while the new one is based entirely on ansible:

-  the number of nodes in the k8s-cluster to be deployed is done through a "configuration/variable file" in the ansible playbook;

-  choosing the version of the packages to be installed (like kubeadm, kubelet ...) is as simple as declaring it in the configuration file;

- the configuration of the vagrant environment is also done through the ansible playbook. When the playbook is played a new Vagrantfile is generated and then used to bring up the VMs; The user does not need to know how a Vagrantfile works to modify the deployment of the cluster.

-  the network configuration of the VMs is setted in bridge mode possibly in a different network interface than the default one of the host machine and the NATed interface brought up by the vagrant engine is disabled. This decision was driven by debug needs when sniffing network traffic on the k8s cluster;

- the access to the machines is done directly through a pre-generated ssh-key without the need of the vagrant command. a to do feature is to automatically generate the key and configure it in the home folder of the user.

### CODE REVIEW

File: tests/ansible_vagrant/kubernetes-playbooks/join-command [hunk 1]
```diff
@@ -0,0 +1 @@
+kubeadm join 192.168.0.10:6443 --token gih27d.voz17y798xogeuov     --discovery-token-ca-cert-hash sha256:58860de2e9a07f7c7fd7ea9e6781a9141689e328d0d83f2d098a3131fffe458a
```
Comment (palexster) [inline tests/ansible_vagrant/kubernetes-playbooks/join-command]: Is this valid for any created cluster?

File: tests/ansible_vagrant/kubernetes-playbooks/vagrant-ssh-key.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# Check the Vagrantfile if you need to change the playbook called during the initial set up
+
+- hosts: all
+  become: true
+  tasks:
+  - name: Set authorized key
+    authorized_key:
+      user: vagrant
+      state: present
+      key: "{{ lookup('file', '/home/aldo/.ssh/vagrant_machines.pub') }}" # Path where the pub key is located
```
Comment (palexster) [inline tests/ansible_vagrant/kubernetes-playbooks/vagrant-ssh-key.yml]: '/home/aldo/.ssh/vagrant_machines.pub' should be inserted among other variables

**Comment by frisso** [comment]
@alacuku Can you please clarify the difference between this vagrant files and the one already linked in the master documentation about K8s ( https://raw.githubusercontent.com/polycube-network/polycube/master/tests/vagrant/pcn-k8s/Vagrantfile )? Thanks!

**Comment by alacuku** [comment]
> @alacuku Can you please clarify the difference between this vagrant files and the one already linked in the master documentation about K8s ( https://raw.githubusercontent.com/polycube-network/polycube/master/tests/vagrant/pcn-k8s/Vagrantfile )? Thanks!
The main difference is the old one does the provision the VMs with an bash script embedded in the vagrant file while the new one is based entirely on ansible:

- the number of nodes in the k8s-cluster to be deployed is done through a "configuration/variable file" in the ansible playbook;

- choosing the version of the packages to be installed (like kubeadm, kubelet ...) is as simple as declaring it in the configuration file;

- the configuration of the vagrant environment is also done through the ansible playbook. When the playbook is played a new Vagrantfile is generated and then used to bring up the VMs.

**Comment by frisso** [comment]
Thanks for the detailed answer.
Two more questions: 
- are the original scripts there to stay, or do you plan to delete them?
- the detailed description you wrote before should be added as a description of the PR, in order to keep this information for the future.

File: tests/ansible_vagrant/Vagrantfile [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    config.vm.define "k8s-master" do |master|
+        master.vm.box = IMAGE_NAME
+        # change the bridge interface to match the one on your host machine
+        master.vm.network "public_network", bridge: BRIDGE_NIC, ip: "192.168.0.10"
+        master.vm.hostname = "k8s-master"
+        master.vm.provision "ansible" do |ansible|
+            # Configures the ssh-key
+            ansible.playbook = "kubernetes-playbooks/vagrant-ssh-key.yml"
+             ansible.extra_vars = {
+                pub_key_path: "/home/aldo/.ssh/vagrant_machines.pub"
```
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: Why are there references to your local file system?

File: tests/ansible_vagrant/Vagrantfile [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    config.vm.define "node-1" do |node|
+        node.vm.box = IMAGE_NAME
+        # change the bridge interface to match the one on your host machine
+        node.vm.network "public_network", bridge: BRIDGE_NIC, ip: "192.168.0.11"
+        node.vm.hostname = "node-1"
+        node.vm.provision "ansible" do |ansible|
+            # Configures the ssh-key
+            ansible.playbook = "kubernetes-playbooks/vagrant-ssh-key.yml"
+            ansible.extra_vars = {
+                pub_key_path: "/home/aldo/.ssh/vagrant_machines.pub"
```
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: here too

File: tests/ansible_vagrant/Vagrantfile [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    config.vm.define "node-2" do |node|
+        node.vm.box = IMAGE_NAME
+        # change the bridge interface to match the one on your host machine
+        node.vm.network "public_network", bridge: BRIDGE_NIC, ip: "192.168.0.12"
+        node.vm.hostname = "node-2"
+        node.vm.provision "ansible" do |ansible|
+            # Configures the ssh-key
+            ansible.playbook = "kubernetes-playbooks/vagrant-ssh-key.yml"
+            ansible.extra_vars = {
+                pub_key_path: "/home/aldo/.ssh/vagrant_machines.pub"
```
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: here too

File: tests/ansible_vagrant/Vagrantfile [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    config.vm.define "node-3" do |node|
+        node.vm.box = IMAGE_NAME
+        # change the bridge interface to match the one on your host machine
+        node.vm.network "public_network", bridge: BRIDGE_NIC, ip: "192.168.0.13"
+        node.vm.hostname = "node-3"
+        node.vm.provision "ansible" do |ansible|
+            # Configures the ssh-key
+            ansible.playbook = "kubernetes-playbooks/vagrant-ssh-key.yml"
+            ansible.extra_vars = {
+                pub_key_path: "/home/aldo/.ssh/vagrant_machines.pub"
```
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: here too

File: tests/ansible_vagrant/Vagrantfile (hunk 3 above)
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: Why are there references to your local file system?

File: tests/ansible_vagrant/Vagrantfile (hunk 4 above)
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: here too

File: tests/ansible_vagrant/Vagrantfile (hunk 5 above)
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: here too

File: tests/ansible_vagrant/Vagrantfile (hunk 6 above)
Comment (michelsciortino) [inline tests/ansible_vagrant/Vagrantfile]: here too

File: tests/ansible_vagrant/kubernetes-playbooks/group_vars/all/vars.yml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#Needed for apt over HTTPS
+apt_packages:
+- apt-transport-https
+- ca-certificates
+- curl
+- gnupg-agent
+- software-properties-common
+
+#path to the public key used to ssh to the machines
+pub_key_path: /home/aldo/.ssh/vagrant_machines.pub
```
Comment (michelsciortino) [inline tests/ansible_vagrant/kubernetes-playbooks/group_vars/all/vars.yml]: here too

File: tests/ansible_vagrant/kubernetes-playbooks/group_vars/all/vars.yml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+#Vagrant file configuration#
+############################
+
+#flavor of the operating system to be used in the VMs
+image_name: "ubuntu/bionic64"
+#the nic on your host to which the VM interface will be bridged.
+bridge_nic: enp1s0f0 
+
+#path to the public key used to ssh to the machines
+pub_key_path: /home/aldo/.ssh/vagrant_machines.pub
```
Comment (michelsciortino) [inline tests/ansible_vagrant/kubernetes-playbooks/group_vars/all/vars.yml]: here too

File: tests/ansible_vagrant/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
+# Ansible playbook
+An ansible playbook to install a simple k8s cluster with a master node and two worker nodes. The CNI plugin is pcn-k8s. Vagrant is used to provide the VMs.
+
+## Requirements
+ * Vagrant 2.2.5 or higher
+ * Ansible 2.8.5 or higher
+
+## Usage
+ * ansible-playbook master-playbook.yml
```
Comment (michelsciortino) [inline tests/ansible_vagrant/README.md]: This should be shown as code by using back-ticks:
 `ansible-playbook master-playbook.yml`

**Comment by michelsciortino** [comment]
@alacuku few considerations

missing:
- a vagrant destroy trap in case of failures running the script

suggestions:
- make it possible to configure the Ansible variables as parameters of the ansible-playbook command, such as `k8s_master_node_ip`, `k8s_worker_node_ips`, `bridge_nic`, `pub_key_path`, `virtual_memory`, `virtual_cpu`)
- rename `virtual_memory` to `virtual_memory_size` and `virtual_cpu` to `virtual_cpus`
- the ssh public key could be not present in the host so the script should check the existence of it and, in case the key is missing, it should propose to generate a new ssh key pair and locate it to the default directory (`~/.ssh`) 
- auto-generate host_vars yaml files in order to manage nodes IPs just in one place (`vars.yml` / command line) + automatically append host IPs and name to /etc/hosts 
- make it possible to choose the number of slave nodes to be automatically created instead of manually create the files and configurations of each of them
- remove every auto-generated file from the repository as they are not needed (i.e. `Vagrantfile` and `join-command` files)
- a more explanatory documentation about how to set the Ansible variables

File: tests/ansible_vagrant/Vagrantfile (hunk 3 above)
Comment (alacuku) [reply tests/ansible_vagrant/Vagrantfile]: the Vagrant file is generated by the playbook. you just need to change the variables in group_vars/all/vars.yml.

Comment (frisso) [reply tests/ansible_vagrant/Vagrantfile]: If it is generated dynamically, why is it in the git repository?

Comment (alacuku) [reply tests/ansible_vagrant/Vagrantfile]: i forgot to put it in gitignore. with the new commit i resolved this issue, and implemented the features and fixes suggested by @michelsciortino.

**Pull request merged**